datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -38,10 +38,15 @@ use datafusion_expr::logical_plan::{
 use datafusion_expr::{col, Expr, ExprSchemable};
 
 /// A map from expression's identifier to tuple including
+///
+/// key == Identifier created with only the current node (& subtree)
+///
+/// values:
 /// - the expression itself (cloned)
 /// - counter
 /// - DataType of this expression.
-type ExprSet = HashMap<Identifier, (Expr, usize, DataType)>;
+/// - symbol used as the identifier in the alias
+type ExprSet = HashMap<Identifier, (Expr, usize, DataType, Identifier)>;
```

Comment:
Turns out that what the IdArray was storing as an identifier, was the Self::expr_identifier(expr) + concatted to the visited stack.  This produced a symbol (a.k.a. a String) used for the alias in the common expression.

That symbol was the only thing meaningful in IdArray. Therefore, I instead moved that symbol into the expr_set. Since the expression set can be keyed to only the `Self::expr_identifier(expr)`, which is available in both the first and second visitor.

In this way, I avoided any use of idx during tree traversal -- whcih required that both visitor traversed the tree in the exact same way.